tpm/tpm.go

```diff
@@ -0,0 +1,199 @@
+	privateArea, publicArea, retErr = tpm2.Seal(rwc, srkHandle, srkPassword, objectPassword, policy, dataToSeal)
+	if retErr != nil {
+		return nil, nil, fmt.Errorf("unable to seal data: %v", err)
+	}
+	glog.Infof("Sealed data: 0x%x\n", privateArea)
+
+	return privateArea, publicArea, nil
+}
+
+// Returns the unsealed data
+func Unseal(tpmPath string, pcr int, srkPassword, objectPassword string, privateArea, publicArea []byte) (data []byte, retErr error) {
+	// Open the TPM
+	rwc, err := tpm2.OpenTPM(tpmPath)
+	if err != nil {
+		return nil, fmt.Errorf("can't open TPM %q: %v", tpmPath, err)
+	}
+	defer func() {
+		if err := rwc.Close(); err != nil {
+			retErr = fmt.Errorf("%v\ncan't close TPM %q: %v", retErr, tpmPath, err)
+		}
+	}()
+
+	// Create the parent key against which to seal the data
+	srkHandle, _, err := tpm2.CreatePrimary(rwc, tpm2.HandleOwner, tpm2.PCRSelection{}, "", srkPassword, srkTemplate)
+	if err != nil {
+		return nil, fmt.Errorf("can't create primary key: %v", err)
+	}
+	defer func() {
+		if err := tpm2.FlushContext(rwc, srkHandle); err != nil {
+			retErr = fmt.Errorf("%v\nunable to flush SRK handle %q: %v", retErr, srkHandle, err)
+		}
+	}()
+	glog.Infof("Created parent key with handle: 0x%x\n", srkHandle)
+
+	// Load the sealed data into the TPM.
+	objectHandle, _, err := tpm2.Load(rwc, srkHandle, srkPassword, publicArea, privateArea)
+	if err != nil {
+		return nil, fmt.Errorf("unable to load data: %v", err)
+	}
+	defer func() {
+		if err := tpm2.FlushContext(rwc, objectHandle); err != nil {
+			retErr = fmt.Errorf("%v\nunable to flush object handle %q: %v", retErr, objectHandle, err)
+		}
+	}()
+	glog.Infof("Loaded sealed data with handle: 0x%x\n", objectHandle)
+
+	// Create the authorization session
+	sessHandle, _, err := policyPCRPasswordSession(rwc, pcr, objectPassword)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get auth session: %v", err)
+	}
+	defer func() {
+		if err := tpm2.FlushContext(rwc, sessHandle); err != nil {
+			retErr = fmt.Errorf("%v\nunable to flush session: %v", retErr, err)
+		}
+	}()
+
+	// Unseal the data
+	unsealedData, err := tpm2.UnsealWithSession(rwc, sessHandle, objectHandle, objectPassword)
+	if err != nil {
+		return nil, fmt.Errorf("unable to Unseal data: %v", err)
+	}
+	return unsealedData, nil
+}
+
+// Returns session handle and policy digest.
+func policyPCRPasswordSession(rwc io.ReadWriteCloser, pcr int, password string) (sessHandle tpmutil.Handle, policy []byte, retErr error) {
+	// FYI, this is not a very secure session.
+	sessHandle, _, err := tpm2.StartAuthSession(
+		rwc,
+		tpm2.HandleNull,  /*tpmKey*/
+		tpm2.HandleNull,  /*bindKey*/
+		make([]byte, 16), /*nonceCaller*/
+		nil,              /*secret*/
+		tpm2.SessionPolicy,
+		tpm2.AlgNull,
+		tpm2.AlgSHA256)
+	if err != nil {
+		return tpm2.HandleNull, nil, fmt.Errorf("unable to start session: %v", err)
+	}
+	defer func() {
+		if sessHandle != tpm2.HandleNull && err != nil {
```

Comment:
Why could sessHandle be Null?
You can always capture its current value as defer func argument

Reply:
Looking at StartAuthSession, it does return nil under several conditions.
Not sure what value I would get by passing sessHandle to defer func since sessHandle is either nil or not nil by the time the defer runs.